core/src/Kokkos_ScratchSpace.hpp

```diff
@@ -31,103 +31,71 @@ static_assert(false,
 
 namespace Kokkos {
 
-/** \brief  Scratch memory space associated with an execution space.
- *
- */
-template <class ExecSpace>
-class ScratchMemorySpace {
+template <typename ExecSpace, typename PointerType>
+class ScratchMemorySpaceBase {
   static_assert(
       is_execution_space<ExecSpace>::value,
-      "Instantiating ScratchMemorySpace on non-execution-space type.");
+      "Instantiating ScratchMemorySpaceBase on non-execution-space type.");
 
  public:
   // Minimal overalignment used by view scratch allocations
   constexpr static int ALIGN = 8;
 
  private:
-  mutable char* m_iter_L0 = nullptr;
-  mutable char* m_iter_L1 = nullptr;
-  char* m_end_L0          = nullptr;
-  char* m_end_L1          = nullptr;
+  mutable PointerType m_iter = nullptr;
+  PointerType m_end          = nullptr;
 
-  mutable int m_multiplier    = 0;
-  mutable int m_offset        = 0;
-  mutable int m_default_level = 0;
-
-#ifdef KOKKOS_ENABLE_DEPRECATED_CODE_4
-  constexpr static int DEFAULT_ALIGNMENT_MASK = ALIGN - 1;
-#endif
+  mutable int m_multiplier = 0;
+  mutable int m_offset     = 0;
 
  public:
   //! Tag this class as a memory space
-  using memory_space    = ScratchMemorySpace<ExecSpace>;
+  using memory_space    = ScratchMemorySpaceBase<ExecSpace, PointerType>;
   using execution_space = ExecSpace;
   //! This execution space preferred device_type
   using device_type = Kokkos::Device<execution_space, memory_space>;
 
   using array_layout = typename ExecSpace::array_layout;
   using size_type    = typename ExecSpace::size_type;
 
-  static constexpr const char* name() { return "ScratchMemorySpace"; }
-
-#ifdef KOKKOS_ENABLE_DEPRECATED_CODE_4
-  // This function is unused
-  template <typename IntType>
-  KOKKOS_DEPRECATED KOKKOS_INLINE_FUNCTION static constexpr IntType align(
-      const IntType& size) {
-    return (size + DEFAULT_ALIGNMENT_MASK) & ~DEFAULT_ALIGNMENT_MASK;
-  }
-#endif
```

Comment:
... this has been lost over the course of changes to this pull request. I am happy add them back in if preferred.